### DISCUSSION TITLE
Circuit Slowness [Issue with v0.23.x]

### DISCUSSION THREAD

**Discussion by wktanghk** [scikit-rf/scikit-rf#764] [Q&A] [answered]
Hi, 

I ran into a problem with my script which is taking more than 8 hours to run. I tracked down the slowness to a line where I tried to access the s-parameter matrix from my Circuit component (called ckt in the example code below)

nw2p = ckt.network

ckt is built from an input S118P file (118-port Sparameter 21-freq points). In this S118P, Port1 & Port2 are each connected to a `Circuit.Port` (50ohm ports). The remaining 116 ports are connected together in pairs. For example, `Port3` connecting to `Port4`; `Port5` connecting to `Port6` ... ... `Port117` connecting to `Port118`. All of these connections are done by building a connection list (`cnx`), which is then used to build ckt.
`ckt = rf.Circuit(cnx)`

A snippet of the `cnx` is pasted at the bottom of this message where P43 & P44 are connected together (this snippet was based on a smalller S84P file, but the concept is the same).

Since I need to do a series of s-matrix manipulation. So I have to call the line  (`nw2p = ckt.network`) many times (~20 times), and each time it is taking 30min. Hence the overall script becomes very slow to complete. I would like to know if this slowness is expected, and if there are more efficient way of doing it. For example, I read that I can also use `network.interconnect` to make the internal connection. Not sure if that would be any faster (?) 

Any advices will be greatly appreciated! Thanks, William

```
[(84-Port Network: 'sp_X1_TxAll_RxSE',  22.0-32.0 GHz, 21 pts, z0=[50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j], 44), (84-Port Network: 'sp_X1_TxAll_RxSE',  22.0-32.0 GHz, 21 pts, z0=[50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j 50.+0.j
 50.+0.j 50.+0.j 50.+0.j], 45)]
```

**Comment by jhillairet** [comment, marked as answer]
Interesting setup. If I'm understanding your problem correctly, the Circuit description could be something like

```python
import skrf as rf
import numpy as np
import time

# Dummy N-ports network - replace ntwk with yours
nports = 300
freq = rf.Frequency(start=1, stop=21, npoints=21)
s = np.random.rand(len(freq), nports, nports)
ntwk = rf.Network(frequency=freq, s=s, name='ntwk')

start = time.time()
# Connexions list 
Port1 = rf.Circuit.Port(freq, 'port1')
Port2 = rf.Circuit.Port(freq, 'port2')

cnx = [[(ntwk, 0), (Port1, 0)],
       [(ntwk, 1), (Port2, 0)],
       *[[(ntwk, idx), (ntwk, idx+1)] for idx in range(2, nports, 2)]
       ]

# Building circuit
cir = rf.Circuit(cnx)

# Building Network
cir_ntwk = cir.network
end = time.time()
print('Circuit network in', end - start)
```
which scales relatively nicely with the number of port (2.7 seconds for 300 ports on my laptop). 

However, I'm not totally sure that `Circuit` is handling properly connexion between the same Network (to be tested)... So, another way (and also much faster) to achieve what you want is probably to use `interconnect` instead:

```python
#%% using innerconnect instead
_ntwk = ntwk.copy()
while _ntwk.nports != 2:
    _ntwk = rf.innerconnect(_ntwk, _ntwk.nports - 2, _ntwk.nports - 1)
# _ntwk is now a 2-port between port0 and port1
innerconnect_ntwk = _ntwk 
```

let me know if your `cir_ntwk` and `innerconnect_ntwk` are equal (seems they are).

**Reply by wktanghk** [reply]
Hi @jhillairet , thank you so much as always! Let me give it a try and get back to you if they are the same and quantify the run time difference. 

It is always impressive to see how you all can quickly capture the gist of the problem in relatively few lines of code.

**Reply by wktanghk** [reply]
[Updated]

Hi @jhillairet,

Thanks again for your script. Strangely, it took a really long time to run with Method #1 (Circuit). I tried it on both my Mac book and Windows workstation. And both used 
Pycharm 2022.1 (Build #PC-221.5080.212 4/12/2012), and scikit-rf 0.23.1.

The line that took a long time to process is (more than 5 hours, and I justed killed it)
cir_ntwk = cir.network

All the lines before that executed in split-second.

Method #2 is faster (90.7sec) but certainly longer than the 2.7sec you observed with Method #1.


Would you have any insight into what may cause this delay? 

[Updated] Fyi - I also tried using Jupyter notebook on a different Windows PC (also with Scikit-rf 0.23.1), but Method #1 was still stuck at the same line for a long time, and I also just killed the process after more than 30min.

Thanks again,
William

**Reply by wktanghk** [reply]
Hi @jhillairet,

One more update- the slowness is specific to v0.23.x.

I reverted back to v0.22.1 and before and it ran quickly (Method #1 with Circuit: 0.39sec; Method #2 with Innerconnect: 77.96sec)

The line that is causing the problem in v0.23.x is:

cir_ntwk = cir.network

Would you recommend we continue to discuss here or should I copy this thread to the Issue forum?

Thanks,
William

**Reply by jhillairet** [reply]
Hum, interesting... Could you share a minimal working example ?

**Reply by wktanghk** [reply]
Actually, it's the exact same code that you posted. (Re-paste here for your reference. Sorry, having some trouble for it to format properly like regular code)

```python
import skrf as rf
import numpy as np
import time

nports = 300
freq = rf.Frequency(start=1, stop=21, npoints=21)
s = np.random.rand(len(freq), nports, nports)
ntwk = rf.Network(frequency=freq, s=s, name='ntwk')

start = time.time()
Port1 = rf.Circuit.Port(freq, 'port1')
Port2 = rf.Circuit.Port(freq, 'port2')

cnx = [[(ntwk, 0), (Port1, 0)], [(ntwk, 1), (Port2, 0)], *[[(ntwk, idx), (ntwk, idx+1)] for idx in range(2, nports, 2)]]

cir = rf.Circuit(cnx)  

cir_ntwk = cir.network ### <-v0.23.x ISSUE with this line
end = time.time()
print('Circuit network in', end - start)

```

**Reply by jhillairet** [reply]
This is indeed very weird, I've tried on two different computers (scikit-rf v0.23.1) and the results are *very* different. 
- The first one is my laptop 8 Go RAM Intel i5 @ 1.6 GHz and SSD drive
- The second one is a Linux server 32 Go Xeon @2.1 GHz and HDD drive
So I would say it could be related to the hard drive read/writing speed (but why?)... What is your drive type?

![image](https://user-images.githubusercontent.com/4642848/196207069-9854950a-2975-4b51-87a1-144e8bf8128e.png)

**Reply by jhillairet** [reply]
OK, I understand. I was using the git development version, not the official 0.23.1. The slowness issue comes from the `s2s` function which is often called, and which is much faster since the PR #738 made by @qmfrf.

So, @wktanghk, if you use the git version of scikit-rf, that should solve your issue and it should be *much* faster. In fact the plot above illustrates the difference before and after the PR #738. 

The current git version should be released in the next weeks as a new conda/pip version anyway.

**Reply by wktanghk** [reply]
Great, than you so much for tracking it down @jhillairet! 

It's my first time downloading a package from GitHub. I downloaded the zip, and installed it in PyCharm. But for some reason, it showed similar problem as the version from pip. Do you have an estimate when the next version would become available? Trying to decide if I should just stay with v0.22.1 for now.

![image](https://user-images.githubusercontent.com/42332848/196602234-c24eab5f-456c-42af-a2b3-bdf05b464709.jpeg)


Thanks,
William

**Reply by Vinc0110** [reply]
No, not with the Python Console, but with the regular terminal (within PyCharm or directly on machine). There you can install the package from the downloaded code:

1. Navigate the scikit-rf base directory (or start terminal when project is already open)
2. Run `python setup.py install`

If you want to get rid of it again, just uninstall with the Python package manager:

1. `pip uninstall scikit-rf`
2. maybe then `pip install scikit-rf` to install the regular scikit-rf package from pypi

**Reply by wktanghk** [reply]
Hi @Vinc0110 , thanks for the recommendation. I uninstalled skrf, and then in cmd, I navigated to where I unzipped the skrf 0.23.1 downloaded from GitHub. And then I typed "python setup.py install". The package installed successfully.

However, it gave similar problem as the version 0.23.1 that I installed from pip before, where the test code was stuck. So now I reverted back to 0.22.1 for now to avoid the issue.

So at this point, I'm leaning towards waiting for the next release so as not to take up too much time from you and @jhillairet. But thanks for teaching me how to install python other than through pip.

Thanks,
William

**Reply by jhillairet** [reply]
OK. For the moment, I'm leaving this as unresolved. When the next version (0.24) will be released, let us know if it fixes your speed issue.

**Reply by jhillairet** [reply]
Solved in scikit-rf 0.24

**Reply by wktanghk** [reply]
Hi, just like to confirm that the issue is solve in scikit-rf 0.24.
Using 0.24, the program is no longer stuck when trying to connect this large matrix using Circuit.

Here are the results for the 2 methods to connect this large matrix:
Method 1: Circuit network in 2.6268346309661865
Method 2: Using Network innerconnect in 104.1707763671875

Thanks!

**Reply by jhillairet** [reply]
Thank you for the feedback.
